### ISSUE TITLE
v2: think about compatibility of v1 complex parsers

### ISSUE THREAD

**Issue by rgerhards** [rsyslog/liblognorm#103]
We need to think if we can emulate them; if not, we need at least document them as no longer present plus a way how to convert the rulebase.

Idea is that we eventually can convert them on the fly into a json prscnf object, but that may not be worth the hassle.

This is to be done late in the v2 cycle, when everything else has stabilized.

**Comment by rgerhards** [comment]
It looks like an automatic conversion is a lot of work, because we would need to old-style parser defs (which we got rid of). So the plan is to make the user convert things, which should be pretty easily given the new user-defined types and "repeat". This is already partly documented in the compatibilty doc.

**Comment by davidelang** [comment]
If we want the distros to accept the new version, there is going to need to be a 
script that can convert the ruleset from the old version to the new version. 
It's better for the liblognorm project to create such a tool rather than having 
distros each create their own.

For the vast majority of parsers, it seems to me that a simple table of "parser 
X extradata becomes parameter Y" conversions that can be applied as a pre-parser 
pass to the ruleset would do the job.

The only cases where this doesn't work are when the extradata is more than one 
string.

looking through the docs

no conversion needed

number
float
hexnumber
kernel-timestamp
whitespace
word
alpha
rest
quoted-stringop-quoted-string
date-iso
time-24hr
time-12hr
duration
date-rfc3164
date-rfc5424
ipv4
ipv6
mac48
cef
checkpoint-lea
cisco-interface-spec
json
cee-syslog

simple parameter conversion

string-to
char-to
char-sep
regex
recursive
descent (two parameters)
interpret (two parameters)

    note, the description of suffixed implies more capabilities than are documented for interpret


complex conversion

tokenized (two parameters, nesting)
iptables (special case, the field name needs to be replaced by .)
suffixed (three parameters, one parameter is a list)
named_suffix (five parameters, one parameter is a list)

nested tokenized is the only item that looks hard.

Am I missing something?

David Lang

**Comment by rgerhards** [comment]
You should note that all regular parser just work - the old syntax is part of the new one and are often preferred even in the new conf because of their compactness. I am just talking about things like recursive which have far better replacements in v2.

**Comment by rgerhards** [comment]
See  also: https://github.com/rgerhards/liblognorm/blob/rger-v2/compatibility-v2

**Comment by davidelang** [comment]
On Tue, 28 Jul 2015, Rainer Gerhards wrote:

> See  also: https://github.com/rgerhards/liblognorm/blob/rger-v2/compatibility-v2

I just found that (I was wondering where this development was going on)

it looks like doc/configuration.rst is not updated for v2 parsers

so the problematic parsers are

iptables
   what's the difference between the v1 and v2 parsers? Isn't iptables just a 
special case of a NameValue parser?

   I use the old iptables quite a bit as a name value parser, I expect others do 
as well.

tokenized
   this needs to be converted to repeat, but I'm not clear on what the symantics 
of repeat are. I see it takes two parameters 'parser' and 'while', parser I 
assume is what is used to read data, but it's not clear what while means.

recursive
   this could be viewed as a special case of descent where the file used is the 
main file. I agree that the resulting parse tree is ugly, especially in the case 
of recursive where every rule is duplicated (I think it's reasonable to only 
allow one level of recursion)

descent
   this has been a poor-man's include with nestig. I think we are going to need 
to have an include (my ruleset is getting ugly to manage as it nears 600 rules)

Given the much better support for named types and nesting of variables that v2 
will have, is it really that ugly to support these?


have the converted create a new type, something like

type=autogen##:%.:include{file='something', warn='depriciated'}
then replace descent in any rules with the specific autogen## type.

and then have it log a message every time it's loaded (unless the user 
explicitly removes the warn= text, in which case they have had their warning)

David Lang

**Comment by davidelang** [comment]
On Tue, 28 Jul 2015, David Lang wrote:

> If we want the distros to accept the new version, there is going to need to 
> be a script that can convert the ruleset from the old version to the new 
> version. It's better for the liblognorm project to create such a tool rather 
> than having distros each create their own.

At the very least, we need to have a script that detects "I can't support this 
config" and refuse to upgrade the package.

David Lang

**Comment by rgerhards** [comment]
Such a script is definitely quite a lot of work (at least for me, not knowing any of the specifics). I think we should wait on the first final version of v2 and then decide on the way forward. I tend to say that I would prefer to re-name liblognorm v2 rather than carry over this legacy because I honestly do not think that anyone would run into any real problems. But there are also ample of other solutions, e.g. I could include the old code together with the new, and just rename the v2 API. I think it clumpsy, but doable. I'll also try to see to get some feedback on the mailing list.

**Comment by rgerhards** [comment]
On the other note: the development is in my private git because it's still far from being finished. Once it approaches completeness, I'll do a PR against the official repo. Note that everything currently in my implementation could change at any instant. That's also the reason why no doc is yet written: it's a moving target. Also, I had a very limited timeframe for core part implementation and would not have managed to go where I am now if I had documented everything as it evolved (and changed again).

Old and new iptables are mostly the same, except that the new one writes data to its own container. The old one can probably simply be emulated by the new one with name "." -- but I haven't looked into that because otherwise I could not finish the core algo in time.

repeat: the "while" part tells what needs to follow so that repeat continues to parse. Simple sample:
    https://github.com/rgerhards/liblognorm/blob/rger-v2/tests/repeat_simple.sh

recursive: IMHO this is a really ugly hack, albeit something that made a lot of sense in v1. User-defined types are a much better solution. Recursive is very sensitive to new rules being added. Keeping the tail is also extremly ugly.

descent: same on tail, again user-defined types are much better. Adding an include once we have the initial version ready should not be any real work (and I agree it makes sense, e.g for shared type definitions). I also think I remember @janmejay said that descent has some load order problems, which we don't have with the new code.

The suffixed parsers I had no time yet to look into. I strongly think they can be cleanly replaced by simple parser arguments, which would be much more powerful.

Why I don't like to support all these things: they are tied into a big, no longer existing config system. I need to remodel that. They also have a lot of parser code, which we don't need for the new style (but granted, we could generate the replacements so we don't need it). I do not like to keep support for them around because I really don't think it's worth it. IMHO only a handful of users use these features. If I know correctly, no major distro has even adopted a version that carries all that into their repository. So to me it boils down to doing a lot of effort for functionality that almost nobody needs, and we would need to support it for all eternity...

**Comment by radu-gheorghe** [comment]
Did anyone actually integrate liblognorm in a major distribution? I wasn't aware of that. Also, my impression was that liblognorm isn't very popular so normally users would fall in two categories:
- early adopters using liblognorm for serious use-cases. They'll know what to do and would probably prefer that work goes into developing v2 instead of keeping backwards compatibility
- people who are just looking to use it and should move to v2 before going into production anyway

I think this is different than rsyslog itself which is widely used and people expect it to be backwards compatible. Though even there one could argue that maybe the old config format should be deprecated so that documentation (rsyslog's main painpoint) can be made much clearer.

**Comment by janmejay** [comment]
I think the best approach would be to have at least a release worth of
overlap between v1 and v2. That way it gives people an easy migration path.
One way I can think of doing this: if we expose another mm-plugin like
mmnormalize, which works with v2, we can leave mmnormalize alone and hence
atleast in rsyslog context allow users to use both mm-plugins at the same
time, hence allowing fully functional rule-by-rule incremental migration.

We can keep v1 for a few months (or an year) to allow everyone enough time
to migrate, then kill mmnormalize (and hence v1 along with it).

After that period, users are expected to have migrated to new mm-plugin,
but that expectation can be set upfront, so it doesn't surprise anyone.

Also, maintainer effort towards v1 in this model should not be very high,
because new parsers we build in v2 need not be made available in v1(so
it'll see no new commits for features, it may get occasional bug-fixes but
that's about it). No new releases, no back-porting of features etc.

Thoughts?

--
Regards,
Janmejay

PS: Please blame the typos in this mail on my phone's uncivilized soft
keyboard sporting it's not-so-smart-assist technology.

**Comment by davidelang** [comment]
On Tue, 28 Jul 2015, Rainer Gerhards wrote:

> On the other note: the development is in my private git because it's still far 
> from being finished. Once it approaches completeness, I'll do a PR against the 
> official repo. Note that everything currently in my implementation could 
> change at any instant. That's also the reason why no doc is yet written: it's 
> a moving target. Also, I had a very limited timeframe for core part 
> implementation and would not have managed to go where I am now if I had 
> documented everything as it evolved (and changed again).

no problem with any of this, it just makes it harder to comment.

> Old and new iptables are mostly the same, except that the new one writes data 
> to its own container. The old one can probably simply be emulated by the new 
> one with name "." -- but I haven't looked into that because otherwise I could 
> not finish the core algo in time.

ok, this one I would do as a one-time upgrade script (sed). I would convert it 
to namevalue not the new iptables as the new iptables looks for specific 
fieldnames and the old one doesn't

> repeat: the "while" part tells what needs to follow so that repeat continues to parse. Simple sample:
>    https://github.com/rgerhards/liblognorm/blob/rger-v2/tests/repeat_simple.sh

this sounds to me like

%a:tokenize:b:c% == %a:repeat{ parser:'%.:c%', while='a' }% (or the full json 
equivalent)

again sed in a one-time upgrade script

with repeat, can the parser be repeat? (i.e. nesting like the example in the 
tokenize doc) that would make the conversion a bit harder, but not impossible.

> recursive: IMHO this is a really ugly hack, albeit something that made a lot 
> of sense in v1. User-defined types are a much better solution. Recursive is 
> very sensitive to new rules being added. Keeping the tail is also extremly 
> ugly.
>
> descent: same on tail, again user-defined types are much better. Adding an 
> include once we have the initial version ready should not be any real work 
> (and I agree it makes sense, e.g for shared type definitions). I also think I 
> remember @janmejay said that descent has some load order problems, which we 
> don't have with the new code.

I agree they are ugly. I also agree that they are better replaced by the new 
types. I'm just worried about the upgrade compatibility. I'm wondering if 
there's a way to progromatically re-write them to be a (very ugly) new type (in 
a one time upgrade script)

so if you have
file1:
rule=:%a;descent:file2:end%

file2:
fule=:%interface:char-to:\x3a%:%ip:ipv4%%end:rest%
fule=:%ip:ipv4%/%port:number%%end:rest%
fule=:%interface:char-to:\x3a%:%ip:ipv4%/%port:number%%end:rest%
fule=:%ip:ipv4%%end:rest%

could file1 be re-written to be:

type=convert01:
type=convert01:%interface:char-to:\x3a%:%ip:ipv4%
type=convert01:%ip:ipv4%/%port:number%
type=convert01:%interface:char-to:\x3a%:%ip:ipv4%/%port:number%
type=convert01:%ip:ipv4%
rule=:%a:convert01%

or if needed, add a! at the beginning of every fieldname in the convert01 type 
(I'm not completely sure of the nesting rules in the new config language)

I agree it would be ugly (especially for recursive), but it would let you 
eliminate the existing (broken) code without breaking users

> The suffixed parsers I had no time yet to look into. I strongly think they can 
> be cleanly replaced by simple parser arguments, which would be much more 
> powerful.

the real value of suffixes is if you can then interpret '1' 'kb' to be '1024', 
otherwise the new user defined types can be used (although there needs to be a 
way to specify a literal and have that literal become a tag

old example from docs:

rule=:reclaimed %eden_reclaimed:suffixed:,:b,kb,mb,gb:number% from eden
Given text "reclaimed 115mb from eden" the above rule would produce:

{
   "eden_reclaimed":
     {
       "value": "115",
       "suffix": "mb"
     }
}

replacement
type=convert02:%value:number% %suffix:literal{text='mb'}%
type=convert02:%value:number% %suffix:literal{text='kb'}%
type=convert02:%value:number% %suffix:literal{text='gb'}%
type=convert02:%value:number% %suffix:literal{text='b'}%
rule=:reclaimed %eden_reclaimed:convert02% from eden

> Why I don't like to support all these things: they are tied into a big, no 
> longer existing config system. I need to remodel that. They also have a lot of 
> parser code, which we don't need for the new style (but granted, we could 
> generate the replacements so we don't need it). I do not like to keep support 
> for them around because I really don't think it's worth it. IMHO only a 
> handful of users use these features. If I know correctly, no major distro has 
> even adopted a version that carries all that into their repository. So to me 
> it boils down to doing a lot of effort for functionality that almost nobody 
> needs, and we would need to support it for all eternity...

If this can be handled by a one-time convert script, would you still be 
as concerned?

David Lang

**Comment by davidelang** [comment]
On Tue, 28 Jul 2015, Janmejay Singh wrote:

> I think the best approach would be to have at least a release worth of
> overlap between v1 and v2. That way it gives people an easy migration path.
> One way I can think of doing this: if we expose another mm-plugin like
> mmnormalize, which works with v2, we can leave mmnormalize alone and hence
> atleast in rsyslog context allow users to use both mm-plugins at the same
> time, hence allowing fully functional rule-by-rule incremental migration.
>
> We can keep v1 for a few months (or an year) to allow everyone enough time
> to migrate, then kill mmnormalize (and hence v1 along with it).
>
> After that period, users are expected to have migrated to new mm-plugin,
> but that expectation can be set upfront, so it doesn't surprise anyone.
>
> Also, maintainer effort towards v1 in this model should not be very high,
> because new parsers we build in v2 need not be made available in v1(so
> it'll see no new commits for features, it may get occasional bug-fixes but
> that's about it). No new releases, no back-porting of features etc.
>
> Thoughts?

there are only a small number of parser types that need to be converted

iptables
recursive
descent
tokenize
suffixed
namedsuffixed

If you aren't using any of these, there's no conversion that must be done 
(although converting to the new JSON for the extradata is a good idea)

keeping the v1 mmnormalize/liblognorm around forever to support this seems like 
a bad idea. Which is why I'm pushing for a conversion script that can re-write 
the old rule file into the new syntax

David Lang

**Comment by davidelang** [comment]
On Tue, 28 Jul 2015, Radu Gheorghe wrote:

> Did anyone actually integrate liblognorm in a major distribution? I wasn't 
> aware of that. Also, my impression was that liblognorm isn't very popular so 
> normally users would fall in two categories:

I believe that mmnormalize and liblognorm are available in many major distros, 
but they don't make use of it by default (in many/most cases it's not installed 
by default)

> - early adopters using liblognorm for serious use-cases. They'll know what to 
> do and would probably prefer that work goes into developing v2 instead of 
> keeping backwards compatibility
>
> - people who are just looking to use it and should move to v2 before going 
> into production anyway
>
> I think this is different than rsyslog itself which is widely used and people 
> expect it to be backwards compatible. Though even there one could argue that 
> maybe the old config format should be deprecated so that documentation 
> (rsyslog's main painpoint) can be made much clearer.

the problem is that an early adopter may have set something up and left, so 
there are people who don't even realize they are using it, and to them it would 
just look like rsyslog broke on an upgrade.

now, this is a fairly narrow window (for most of the problem parsers, iptables 
has been around for years), but remember that we routinely tell people to 
upgrade to the current version of syslog, and doing that will drag in the newest 
liblognorm as well.

David Lang

**Comment by rgerhards** [comment]
I only have a couple of hours left on v2 before I leave for vacation so I
can only comment briefly. David, you need to read up on those problem
parsers. It's much more than just changing syntax. They require this
tail/rest field which imho was a work around to create the functionality in
v1. So we need to carry over the old engine - v2 is totally different even
from a data structure point of view. No easy conversion, just believe me
for now.

A solution is probably possible following @janmejay I think I can include
the old code, add a new "version=2" header to rule base files and call the
old code if it is not present. So the user would by default use the old
stuff and need to opt in to the new one. We could plahse this out after
some time.

Rainer

Sent from phone, thus brief.
Am 29.07.2015 09:09 schrieb "davidelang" <notifications@github.com>:

**Comment by davidelang** [comment]
On Wed, 29 Jul 2015, Rainer Gerhards wrote:

> I only have a couple of hours left on v2 before I leave for vacation so I
> can only comment briefly.

worry about this after vacation, especially if it means leaving something we can 
beat on in the meantime :-)

> David, you need to read up on those problem
> parsers. It's much more than just changing syntax. They require this
> tail/rest field which imho was a work around to create the functionality in
> v1. So we need to carry over the old engine - v2 is totally different even
> from a data structure point of view. No easy conversion, just believe me
> for now.

I used descent for a bit before you implemented the cisco parser, but I think 
the tail/rest just isn't needed using the new type system. once we have 
something that's minimally useful I can test conversions

> A solution is probably possible following @janmejay I think I can include
> the old code, add a new "version=2" header to rule base files and call the
> old code if it is not present. So the user would by default use the old
> stuff and need to opt in to the new one. We could plahse this out after
> some time.

that would be far better than separate packages, but a single-shot conversion 
script would be better (and would let us convert the syntax of the simple types 
as well), if it can work.

David Lang

**Comment by Whissi** [comment]
:+1: to support both. Everything which was public must stay. You need very good reasons to break it which I don't see here. Don't start judging on how many people maybe using v1 syntax already. If it was every released in a final version, deal with it.

The solution from @janmejay with the header sounds good. If you implement this very clean this should help us with future updates, too.

Remember to set the appropriate soname if you are going to break v1.

**Comment by rgerhards** [comment]
I'll add v1 as a subset, because I think it gets very tricky to build packages for rsyslog where one module depends on the old version and another one on the new. I also think it hinders adoption more than that it helps if a user needs to decide which version to bind to. If we really need both old any new, we need to ship both codes, no way around this. I still think its bloated, but so it is. As a side-effect it eases some edges for me as well (e.g. the log structure analyzer does not immediately need to be converted to the v2 subsystem). It's not much code that will be shipped and the v1 code that's then present will not be maintained in any case (except for security updates). So it's probably just a few hours for making sure we call the old stuff whenever the new one is not explicitely requested.

**Comment by radu-gheorghe** [comment]
-1 for having any backwards compatibility in v2, unless it directly benefits v2. Here's my thinking:
- it's a major version. You are allowed to make backwards-incompatible changes by definition. Apps using v1 wanting to migrate to v2 need to be aware of that and update their stuff
- rsyslog is one of those apps. Maybe using v2 in mmnormalize justifies v9. But that's a separate topic from this one that needs to be handled separately IMHO
- if I understand correctly from Rainer, v2 works *very* differently so making it backwards-compatible with v1 requires work, adds complexity and slows it down. I don't think we want that for a young project that we want to make popular (i.e. make it the best it can be so people start using and contributing to it, which will keep the ball rolling). I think we have time to worry about backwards compatibility down the road when the project becomes more stable and mature

**Comment by rgerhards** [comment]
ok, long story short: the v1 engine is now merged into v2 and will be used whenever a rulebase file does not contain "version=2" as the very first line of the rulebase.

@radu-gheorghe you are right, the v1 and v2 engines are totally different. At its core, v1 uses a kind of patricia tree, whereas v2 uses a DAG with disconnected components. Many other things are different as well.

**Comment by davidelang** [comment]
On Wed, 29 Jul 2015, Radu Gheorghe wrote:

> -1 for having any backwards compatibility in v2, unless it directly benefits v2. Here's my thinking:
> - it's a major version. You are allowed to make backwards-incompatible changes by definition. Apps using v1 wanting to migrate to v2 need to be aware of that and update their stuff
> - rsyslog is one of those apps. Maybe using v2 in mmnormalize justifies v9. But that's a separate topic from this one that needs to be handled separately IMHO
> - if I understand correctly from Rainer, v2 works *very* differently so making it backwards-compatible with v1 requires work, adds complexity and slows it down. I don't think we want that for a young project that we want to make popular (i.e. make it the best it can be so people start using and contributing to it, which will keep the ball rolling). I think we have time to worry about backwards compatibility down the road when the project becomes more stable and mature

Please look back through this thread for one of the first big posts that I made. 
I list all the different parsers that exist and there are only ~5 that are a 
problem. The rest will have backwards compatibility.

liblognorm has been starting to receive outside contributions over the last 6-9 
months, and unfortunantly it's some of those contributions that are the problem 
ones.

David Lang

**Comment by rgerhards** [comment]
side note to @radu-gheorghe the relevant parts of the v1 engine are now a part of v2. The effect on v2 is very low: it's actually one if condition per normalize call: if it is a v1 engine, processing is deferred to the v1 part. That's it. Code maintainance should also not be a real issue, as I do not intend to apply any changes to v1, so this part can be considered read-only.

@davidelang I wouldn't object if you want to contribute a conversion script, but I think it is a real lot of work, because you need to do all the recursion in config param processing that e.g. tokenize does. I still doubt it's well spent time, as IMHO only very, very few users use these construct in very, very few places. I intend, however, to well-document in the release notes which conversions need to be done. But that only makes sense when the new format is totally stable. Currently its just PoC (albeit pretty stabilizing).

**Comment by rgerhards** [comment]
another closing note: the more than side-effect of this is that I can make v2 available pretty soon as part of rsyslog, simply because by default everything is handled by the v1 engine. So users can play with it. Actually, one can build liblognorm from my repository and use it together with current rsyslog. Switching back and forth is easy by just adding or removing line 1 as "version=2". That's a nice capability to have.

**Comment by rgerhards** [comment]
one hopefully final note going back through the issue postings:

@davidelang I think we can cleanly replaced suffixed with something like

%byteswritten:integer{"type":"unsigned", "size_suffixes":true}%

That's what I meant with parser parameters. Note that the above is hypothetical.

On repeat: it's a regular parser in v2, so it can be recursively nested as deeply as needed.

Hope that clarifies.

**Comment by crackytsi** [comment]
Ok, topic seems to be solved :) Maybe v1 should be marked as deprecated, to reduce complexity in the future.

Just from someone who is not as active as you: 
I've got about 100 normalization rules (that I want to publish as soon as I solve my "rest-parser" issue) and the only thing I'm using from this "keywords" is the prefix ;)

regards, Chris

**Comment by davidelang** [comment]
this script converts tokenized and iptables items to the closest v2 equivalent

#!/bin/bash
(
echo "version=2"
sed -e 's/%\([^:^%]*\):tokenized:\([^:]*\):\([^%]*\)%/%\1:repeat{ "parser": {"type":"\3", "name":"."}, "while":{"type":"literal", "text":"\2"} }%/g' -e 's/repeat{ "parser": {"type":"\([^:^%^"]*\):\([^"]*\)"/repeat{ "parser": {"type":"\1", "extradata":"\2"/g' -e 's/%\([^:^%]*\):iptables%/%.:name-value-list%/' rsyslog.rulebase
) >rsyslog.rulebase_v2

besides the tokenized vs repeat bug I just files, this has a problem that name-value-list is pickier than v1 iptables was. v1 iptables allowed anything but ' ' and '=' in a name, name-value-list allows alphanumeric ' ' '-' '_' I had some logs that had names like free(file)=value in them.
